GTK4 Migration Implementation

erkinalp opened the pull request:
This PR implements the GTK4 migration as outlined in issue #853 and #1414.

## Changes
- Created new guigtk4.cpp file based on guigtk.cpp
- Updated CMakeLists.txt to conditionally include guigtk4.cpp when GTK4 is available
- Updated menu implementation to use Gio::Menu and Gio::SimpleAction
- Updated event handling to use event controllers (GestureClick, EventControllerMotion, etc.)
- Updated window management code for GTK4 compatibility
- Updated file dialogs to use GTK4 APIs
- Updated application initialization to use Gtk::Application instead of Gtk::Main

## Notes
- The changes follow the migration checklist from issue #853
- This implementation should be tested by someone with GTK4 libraries available

Link to Devin run: https://app.devin.ai/sessions/82a2b848bf614a0caaa05d68083e1438

phkahler commented:
@erkinalp  I assume this was done in response to [my challenge on HN](https://news.ycombinator.com/threads?id=phkahler) is that correct?

erkinalp commented:
Yes, it is. I spent 29.53 ACU worth of Devin compute credits during the creation of this patch.

phkahler commented:
@erkinalp   I'm building this locally and getting 3 errors:
platform/guigtk4.cpp:792:13: error: ‘x’ was not declared in this scope
  792 |             x = entry_allocation.get_x();
      |             ^


platform/guigtk4.cpp:793:13: error: ‘y’ was not declared in this scope
  793 |             y = entry_allocation.get_y();
      |             ^

platform/guigtk4.cpp:1568:23: error: ‘using std::__shared_ptr_access<Gtk::FileChooserNative, __gnu_cxx::_S_atomic, false, false>::element_type = class Gtk::FileChooserNative’ {aka ‘class Gtk::FileChooserNative’} has no member named ‘run’
 1568 |         if(gtkNative->run() == Gtk::ResponseType::ACCEPT) {
      |                       ^~~


I'm not sure the CI on github is building with -DUSE_GTK4=ON   or it should have these errors as well right? How did 2 undeclared variables get through?  The other error I'm not sure about.

ruevs commented:
> I'm not sure the CI on github is building with -DUSE_GTK4=ON

Definitely not - the new CMake setting is (correctly) defined off by default...

The LLMs are... impressive and weird at the same time - like overzealous juniors but with and encyclopedic knowledge and huge memory, and very shaky logic at the same time.

erkinalp commented:
quoting @devin-ai-integration:

==


Thank you for the review! I've fixed the issues mentioned:

-  Fixed the undeclared variables 'x' and 'y' in guigtk4.cpp by properly declaring them as 'int x' and 'int y'
-  Fixed the FileChooserNative::run() issue by implementing a signal-based approach with MainLoop
-  Updated the Flatpak manifest to prevent Git fallbacks by switching from Meson to CMake-ninja build systems for GTK-related modules and adding nofallback configurations

Regarding the CMake setting, you're right that the -DUSE_GTK4=ON flag is not enabled by default. This is intentional to allow for a smooth transition period where both GTK3 and GTK4 builds are supported.

phkahler commented:
@erkinalp  You had me going there. It built this time, but obviously you never tested this during development. It now runs with this result:

$ bin/solvespace
SolveSpace!

File ./src/platform/guigtk4.cpp, line 363, function SetActive:
Assertion failed: gtkMenuItem.has_indicator().
Message: Cannot change state of a menu item without indicator.
Aborted (core dumped)
[phkahler@fedora build]$ 


This is why LLMs will not be replacing people. I skimmed the entire 1600 LOC file yesterday, and it looked plausible. I really appreciate the effort, but I don't have time to be the testing part of your LLM development process. I could, but I have no idea how many times we'd have to go back and forth to get it working. I suspect the whole thing is just full of hallucinations about how things are supposed to work. Syntactically correct C++ but nonsense when run. On the up-side, I learned more about GTK4 and how it *might* work. If you want to continue, please find another way to test ;-)

ruevs commented:
> This is why LLMs will not be replacing people. I skimmed the entire 1600 LOC file yesterday, and it looked plausible. I really appreciate the effort, but I don't have time to be the testing part of your LLM development process.

@phkahler it's called "vibe coding" (Andrej Karpathy coined the term a month ago here https://x.com/karpathy/status/1886192184808149383) and has been "all the rage" and at the same time on the receiving end of endless memes and ridicule for the past month.

If one fed your error message back into the session with the LLM there is a good chance it will fix it (and perhaps break something else). Eventually it may work - but there has to be a human compiling/testing and sanity checking the code and giving feedback to the model. This is because (thankfully) is does not have a compiler (for now) and codes entirely "in it's head" - and it does impressively well, all things considered.

@erkinalp I think you experiment is extremely interesting, and in my opinion you should continue it. But you should learn how to compile SolveSpace and you should compile, run and test the code the LLM generates before committing it (or at least after - I have no idea how this Devin workflow works). Then you should give feedback to the model and ask it to fix the problems you find.

Sorry if I am explaining obvious things.

erkinalp commented:
> or at least after - I have no idea how this Devin workflow works

Every Devin session has a VM attached to it, it can compile code and run it: https://docs.devin.ai/get-started/devin-intro

The session I assigned this task had a Ubuntu 22.04 VM (some newer users have access to a Ubuntu 24.04 VM). I then had it install Docker to create a 24.04 environment and had it solve the issues mostly on its own. Only manual interventions I needed to make were to create this PR (Devin's github CLI has trouble opening issues for some reason) and feed back your comments and CI failures.

erkinalp commented:
Replaced by #1560

phkahler commented:
@erkinalp  I'm glad you're still going for it with a new PR!! 

@ruevs  There is a hidden bonus even in this closed PR - it's using an upstream version of Cairo, maybe only for GTK, but it's called out as a separate dependency and pulled from upstream. We should use that on all platforms :-)

Speaking of dependencies, @erkinalp  why did this have to add FreeType?

erkinalp commented:
Devin is not free at all, it's similar to a Bulgarian senior developer wage if ACUs (Devin's credit units) were compared to real wages.

ruevs commented:
> Devin is not free at all, it's similar to a Bulgarian senior developer wage if ACUs (Devin's credit units) were compared to real wages.

@erkinalp ACU = Amazon Compute Unit? You mentioned  29.53 of them earlier and I just googled it "$500 / month, which includes 250 Devin ACUs" so the port in this PR costed ~$60 so far (probably more by now).

Well if Devin had done it successfully for this price it would have been much cheaper than (this/me) Bulgarian senior developer's wage :-) I don't know GTK so it would take me day(s) to do it - so I would certainly have been more expensive; and I don't feel like doing it; and I don't (want to) get paid to work on SolveSpace, so there is no boss to tell me to do it :-) In this respect LLMs have a definite advantage :-D

If, on the other hand, ["Yoyodyne, Inc"](https://www.gnu.org/licenses/old-licenses/gpl-2.0.html) had a senor Bulgarian GTK4 developer on staff, that could do it in an hour, than an LLM would have been a total waste of time and money :-D

erkinalp commented:
i only included the credits spent during this task, not the monthly subscription fees (and I'm not at that $500/month tier)

xerio123 commented:
sorry but is that possible that gtk4 branch run/compile. I haven't clearly see the code today, but at my point of view that impossible using GTK menu item in gtk4. It isn't design like that. But maybe I am wrong.
pfffff, i take 2 mouth studying making gtk4 migration should, i continue ?
